### ISSUE TITLE
Regression when loading NuGet in MSBuild tasks

### ISSUE THREAD

**Issue by ericstj** [dotnet/msbuild#6186]
### Issue Description
https://github.com/dotnet/msbuild/pull/6126 pins NuGet loads from tasks to use the one in the SDK directly regardless of the version referenced by tasks.  This will unnaturally pin NuGet to the SDK version which will cause issues when tasks need to move ahead either for testing or for general functionality.  In the past we've been very careful about pinning NuGet.

Additionally, the closure of NuGet's assemblies is not loaded from ALC.  This leads to type unification problems, since the first load from MSBuild will cause NuGet to bind to these assemblies in MSBuild, then a task which runs later will bind to it's local copy (even if it is the exact same version).  https://github.com/dotnet/runtime/pull/48462#issuecomment-782854316
```
.packages/microsoft.dotnet.build.tasks.packaging/6.0.0-beta.21120.1/build/Packaging.targets(747,5): error MSB4018: (NETCORE_ENGINEERING_TELEMETRY=Build) The "GenerateRuntimeDependencies" task failed unexpectedly.
System.MissingMethodException: Method not found: 'Void NuGet.RuntimeModel.JsonObjectWriter..ctor(Newtonsoft.Json.JsonWriter)'.
   at Microsoft.DotNet.Build.Tasks.Packaging.NuGetUtility.WriteRuntimeGraph(String filePath, RuntimeGraph runtimeGraph)
   at Microsoft.DotNet.Build.Tasks.Packaging.GenerateRuntimeDependencies.Execute() in /_/src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeDependencies.cs:line 144
   at Microsoft.Build.BackEnd.TaskExecutionHost.Microsoft.Build.BackEnd.ITaskExecutionHost.Execute()
   at Microsoft.Build.BackEnd.TaskBuilder.ExecuteInstantiatedTask(ITaskExecutionHost taskExecutionHost, TaskLoggingContext taskLoggingContext, TaskHost taskHost, ItemBucket bucket, TaskExecutionMode howToExecuteTask)
```

cc @ViktorHofer 

### Steps to Reproduce
1. Run a task which uses both NuGet and Newtonsoft.JSON and calls NuGet APIs which use Newtonsoft.JSON types.  EG: https://github.com/dotnet/arcade/blob/ca7fab569267ed3bc73360882d652d119aae5653/src/Microsoft.DotNet.Build.Tasks.Packaging/src/GenerateRuntimeDependencies.cs

### Expected Behavior
Task runs to completion.

### Actual Behavior
Task fails with MissingMethodException since the calling task assembly and NuGet don't agree on the assembly that Newtonsoft.Json.JsonWriter lives in.

### Analysis
I think we need to rethink the solution for https://github.com/dotnet/msbuild/pull/6126.  I don't think the host should be pinning assemblies which don't need to unify.

It should be *possible* for tasks to defer to MSBuild's copy of assemblies if that works well for them.  In those cases the task should omit the entire assembly closure graph for those assemblies from their task layout, and they should ensure that they reference a version that is <= the one in MSBuild.  As part of the performance investigation/work we should go and talk to the folks that are bundling NuGet and ask them if they are OK with making this change to improve performance.

### Versions & Configurations
dotnet sdk 6.0.100-preview.2.21118.7

Binlog from when building the coreclr packages in dotnet/runtime which shows the `MissingMethodException`: 
[msbuild.zip](https://github.com/dotnet/msbuild/files/6031275/msbuild.zip)

**Comment by ericstj** [comment]
cc @marcpopMSFT @ladipro @brianrob @stephentoub @ViktorHofer 
FYI: This might be preview2 blocking

**Comment by marcpopMSFT** [comment]
@ericstj do you mean net6 preview 2 or 16.10 preview2 as the former has two days and the latter has a month?  If the former, I'd recommend that Brian just back out the previous change and we get approval to merge the dependency flow once this gets into the SDK repo.  Note that 16.10-preview1 has this in it I believe and it's too late for that.

**Comment by ericstj** [comment]
I was just thinking the latter.

**Comment by brianrob** [comment]
Thanks @ericstj.  In that case, I'll abandon the straight revert (#6195), and go with the revert and fix (#6189)..

**Comment by ericstj** [comment]
Actually I might have confused this as Marc was asking me something different in chat. If possible let's add revert to preview2.

**Comment by marcpopMSFT** [comment]
To clarify @brianrob, I had an offline conversation with Eric because of the short timeframe. He wants this fixed in preview 2 for which we have only a couple more days so I suggest still proceeding with the straight revert for net6 preview 2 and get the full fix in for preview3/16.10 preview 3.

**Comment by brianrob** [comment]
Yup, @ericstj and I are having the same conversation right now.  The straight revert looks ready.  I just don't have merge permissions for it.
